libs/common/RespReadUtils.cs

```diff
@@ -4,156 +4,372 @@
-        /// <param name="end"></param>
-        /// <returns></returns>
-        public static bool ReadHeaderLength(out int len, ref byte* ptr, byte* end)
+        /// <param name="ptr">String to try reading sign from.</param>
+        /// <param name="negative">Whether the sign is '-'.</param>
+        /// <returns>True if either '+' or '-' was found, false otherwise.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryReadSign(byte* ptr, out bool negative)
         {
-            len = -1;
-            if (ptr + 3 >= end)
-                return false;
+            negative = (*ptr == '-');
+            return negative || (*ptr == '+');
+        }
 
-            Debug.Assert(*ptr == '$');
-            ptr++;
-            bool neg = *ptr == '-';
-            int ksize = *ptr++ - '0';
-            while (*ptr != '\r')
+        /// <summary>
+        /// Tries to read an unsigned 64-bit integer from a given ASCII-encoded input stream.
+        /// </summary>
+        /// <param name="ptr">Pointer to the beginning of the ASCII encoded input string.</param>
+        /// <param name="end">The end of the string to parse.</param>
+        /// <param name="value">If parsing was successful, contains the parsed ulong value.</param>
+        /// <param name="bytesRead">If parsing was successful, contains the number of bytes that were parsed.</param>
+        /// <returns>
+        /// True if a ulong was successfully parsed, false if the input string did not start with
+        /// a valid integer or the end of the string was reached before finishing parsing.
+        /// </returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool TryReadUlong(ref byte* ptr, byte* end, out ulong value, out ulong bytesRead)
+        {
+            bytesRead = 0;
+            value = 0;
+            var readHead = ptr;
+
+            // Fast path for the first 19 digits.
+            // NOTE: UINT64 overflows can only happen on digit 20 or later (if integer contains leading zeros).
+            var fastPathEnd = ptr + 19;
+            while (readHead < fastPathEnd)
             {
-                Debug.Assert(*ptr >= '0' && *ptr <= '9');
-                ksize = ksize * 10 + *ptr++ - '0';
-                if (ptr >= end)
+                if (readHead > end)
+                {
                     return false;
+                }
+
+                var nextDigit = (uint)(*readHead - '0');
+                if (nextDigit > 9)
+                {
+                    goto Done;
+                }
+
+                value = (10 * value) + nextDigit;
+
+                readHead++;
             }
 
-            ptr += 2;
-            if (ptr > end)
-                return false;
-            Debug.Assert(*(ptr - 2) == '\r');
-            Debug.Assert(*(ptr - 1) == '\n');
-            len = neg ? -ksize : ksize;
+            // Parse remaining digits, while checking for overflows.
```

Comment:
Exactly! The current code allows leading zeros and therefore inputs can be more than twenty digits. We just loop on the digits remaining digits, but in the normal case we will just exit this loop right away (either because we don't have any remaining digits, or because the string did not have leading zeros, and this is the last digit).